runtime/bcutil/defineclass.c

```diff
@@ -134,6 +140,23 @@ internalDefineClass(
 		romClass = createROMClassFromClassFile(vmThread, &loadData, localBuffer);
 	}
 	if (romClass) {
+		/* Host class can only be set for anonymous classes, which are defined by Unsafe.defineAnonymousClass.
+		 * For other cases, host class is set to NULL.
+		 */
+		if ((NULL != hostClass) && (J2SE_VERSION(vm) >= J2SE_19)) {
+			J9ROMClass *hostROMClass = hostClass->romClass;
+			/* This error-check should only be done for anonymous classes. */
+			Trc_BCU_Assert_True(isAnonFlagSet);
+			/* From Java 9 and onwards, set IllegalArgumentException when host class and anonymous class have different packages. */
+			if (!hasSamePackageName(romClass, hostROMClass)) {
+				PORT_ACCESS_FROM_JAVAVM(vm);
+				omrthread_monitor_exit(vm->classTableMutex);
+				setIllegalArgumentExceptionHostClassAnonClassHaveDifferentPackages(vmThread, romClass, hostROMClass);
+				j9mem_free_memory(romClass);
```

Comment:
The romClass isn't allocated via j9mem, its allocated in a memory segment (segment.c). Two choices, either find the memory segment from classloader and use freeMemorySegment() to free it (which should work because anonymous rom classes get their own memory segment), or, move the package check even earlier into ROMClassBuilder.cpp and avoid creating the rom class entirely.

Reply:
added code to clean J9ROMClass. 

I think it should be fine if we don't free J9ROMClass immediately since the ROM class will be inaccessible without a J9Class. The ROM class will be cleaned during shutdown when memory segments are freed.

Reply:
yes we could leave the rom class segment hanging around since this error condition should never happen anyway, but having what is essentially a memory leak doesn't seem like good practice.